### PR TITLE
Adds bounding boxes conversion

### LINKED ISSUE #2687: Bounding Boxes Conversions
## 🚀 Feature

Simple feature request, we have a lot of useful bounding box operations, in torchvision.ops.boxes.py.

## Motivation

Most data format come in either these two, VOC and COCO for object detection.
If we can provide a utility to convert it, it might ease the process from the user's side, just like how IoU and NMS did.

## Pitch

Provide 2 functions, (please name them better)
```
def xywh_from_xyxy(torch_tensor : Tensor[x1 y1 x2 y2]):
   # Add the 2 dimensions of (y2-y1), (x2 - x1) to x2 and y2.

def xyxy_from_xywh(torch_tensor : Tensor[x y w h]):
   # x2 = x + w
   # y2 = y + h
   # Replace these 2 to x2 and y2.
```
## Alternatives

I am unsure of any, for now, we can leave this to the user as well, but this is just utility no forcing.

## Additional context

These are very commonly used in most repositories. Supporting natively like NMS and IoU provides a standard.
Again this is a utility so not a big feature or addition which we need, might help to simplify detection models. I'm unsure in segmentation models.

I can most probably submit a PR for this if it is good to go.

cc @pmeier

### PR DESCRIPTION
**Pull request by oke-aditya**
Closes #2687

- [x] Added code
- [x] Added documentation
- [x] Added tests

I have added as per the Issue, two utility function to convert boxes to pascal VOC format (x1 y1 x2 y2).

The tests convert boxes to other format and vice-versa. This ensures that both operations are identical and can be interchangeably used. Tests passed locally. 

This is ready for review. Do let me know!

cc @pmeier

### CODE REVIEW

File: test/test_ops.py [hunk 1]
```diff
@@ -647,6 +647,48 @@ def test_convert_boxes_to_roi_format(self):
...
+        # Simple test convert boxes to xywh and back. Make sure they are same.
+        # box_tensor is in x1 y1 x2 y2 format.
+        box_tensor = torch.tensor([[0, 0, 100, 100], [0, 0, 0, 0],
+                                  [10, 15, 30, 35], [23, 35, 93, 95]], dtype=torch.float)
+        exp_xywh = torch.tensor([[0, 0, 100, 100], [0, 0, 0, 0],
+                                [10, 15, 20, 20], [23, 35, 70, 60]], dtype=torch.float)
+
+        box_xywh = ops.box_xyxy_to_xywh(box_tensor)
+        # print(box_xywh)
+        # print(box_xywh.shape)
```
Comment (pmeier) [inline test/test_ops.py:661]:
```suggestion
```

File: test/test_ops.py [hunk 2]
```diff
@@ -647,6 +647,48 @@ def test_convert_boxes_to_roi_format(self):
...
+        assert(exp_xywh.size() == torch.Size([4, 4]))
+        assert(exp_xywh.dtype == box_tensor.dtype)
+        assert(torch.all(torch.eq(box_xywh, exp_xywh)).item() is True)
```
Comment (pmeier) [inline test/test_ops.py:664]: For the purpose of learning:

1. `assert` is a [statement](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement) and not a function. Thus, the outer parentheses are not needed
2. The `assert expression` already tests if `bool(expression) is True`. Thus, the prefixed `is True` on a boolean is not needed.

You can apply this in multiple places below.

```suggestion
        assert exp_xywh.size() == torch.Size([4, 4])
        assert exp_xywh.dtype == box_tensor.dtype
        assert torch.all(torch.eq(box_xywh, exp_xywh)).item()
```

File: torchvision/ops/boxes.py [hunk 3]
```diff
@@ -133,6 +133,88 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
     return clipped_boxes.reshape(boxes.shape)
 
 
+def box_cxcywh_to_xyxy(boxes: Tensor) -> Tensor:
+    """
+    Converts bounding boxes from (cx, cy, w, h) format to (x1, y1, x2, y2) format.
```
Comment (pmeier) [inline torchvision/ops/boxes.py:138]: I'm OK with using the short form `cx` and `cy` in the function name, but I think we should add an explanation in the doc string. Could you add a short description what each value means?

File: torchvision/ops/boxes.py [hunk 4]
```diff
@@ -133,6 +133,88 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
...
+    # We need to change all 4 of them so some temporary variable is needed.
+    x1 = boxes[:, 0] - (0.5 * boxes[:, 2])
+    y1 = boxes[:, 1] - (0.5 * boxes[:, 3])
+    x2 = boxes[:, 0] + (0.5 * boxes[:, 2])
+    y2 = boxes[:, 1] + (0.5 * boxes[:, 3])
+
+    # boxes[:, 0] = x1
+    # boxes[:, 1] = y1
+    # boxes[:, 2] = x2
+    # boxes[:, 3] = y2
```
Comment (pmeier) [inline torchvision/ops/boxes.py:155]:
```suggestion
```

File: torchvision/ops/boxes.py [hunk 5]
```diff
@@ -133,6 +133,88 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
...
+
+    cx = (boxes[:, 0] + boxes[:, 2]) / 2
+    cy = (boxes[:, 1] + boxes[:, 3]) / 2
+    w = boxes[:, 2] - boxes[:, 0]
+    h = boxes[:, 3] - boxes[:, 1]
+
+    # boxes[:, 0] = cx
+    # boxes[:, 1] = cy
+    # boxes[:, 2] = w
+    # boxes[:, 3] = h
```
Comment (pmeier) [inline torchvision/ops/boxes.py:181]:
```suggestion
```

**Review by pmeier — COMMENTED** [review]
Thanks @oke-aditya. Few comments below. Additionally:

1. For completeness: shouldn't we also have `box_cxcywh_to_xywh` and `box_xywh_to_cycywh`? I know [you advocated against it](https://github.com/pytorch/vision/issues/2687#issuecomment-699032024), but I think we should discuss this. @fmassa?
2. Why did you add the `pytorch-sphinx-theme` as submodule? I'm pretty sure we shouldn't do that, as the documentation is not build here.

**Comment by oke-aditya** [comment]
I will do the suggested changes.

Also,
1. I advised against it as we can simply cascade these two operations. `box_xywh_to_xyxy(box_cxcy_to_xywh)` and obtain the required result if needed. This is something which I guess we can leave or have a function which exactly does this.

2. Very sorry for sphinx docs, I committed and pushed by mistake. I have removed them.

**Comment by pmeier** [comment]
> I advised against it as we can simply cascade these two operations.

True, but in that case I would ask, why we went for this set of two representations and not some other way. In general, IMO for conversion functions it is always a good idea to have a "core" representation and perform all other conversions only to and from this. Since we only have 3 different representations here, I think we should simply implement them all.

> Very sorry for sphinx docs

Don't be. That is why we have code review :wink:

**Comment by oke-aditya** [comment]
The reason why this of representations is the detection models is torchvision accept x y x y format.
Also why only xywh to xyxy and cxcywh to xyxy is because only these two I could find which were being used, do let me know if other are needed!

I agree that we should provide generic functions. Right now we have only 2 different representations (cxcywh xywh) to provide.
Later if these increase, we would be bound to provide other conversions for consistency.

Hence, building minimal stuff that can do the job was the plan.

I would be happy to add these functions for interconvertability but let @fmassa have a thought.
Both the sides have fair point! It's just choice and design we decide

File: test/test_ops.py (hunk 1 above)
Comment (oke-aditya) [reply test/test_ops.py:661]: Solved 👍

File: test/test_ops.py (hunk 2 above)
Comment (oke-aditya) [reply test/test_ops.py:664]: Absolutely, I corrected it.

File: torchvision/ops/boxes.py (hunk 4 above)
Comment (oke-aditya) [reply torchvision/ops/boxes.py:155]: removed 👍

File: torchvision/ops/boxes.py (hunk 5 above)
Comment (oke-aditya) [reply torchvision/ops/boxes.py:181]: Removed

File: torchvision/ops/boxes.py (hunk 3 above)
Comment (oke-aditya) [reply torchvision/ops/boxes.py:138]: Can you suggest how it should be, a bit confused here.

File: test/test_ops.py [hunk 6]
```diff
@@ -647,6 +647,46 @@ def test_convert_boxes_to_roi_format(self):
...
+    def test_bbox_xywh(self):
+        # Simple test convert boxes to xywh and back. Make sure they are same.
+        # box_tensor is in x1 y1 x2 y2 format.
+        box_tensor = torch.tensor([[0, 0, 100, 100], [0, 0, 0, 0],
+                                  [10, 15, 30, 35], [23, 35, 93, 95]], dtype=torch.float)
+        exp_xywh = torch.tensor([[0, 0, 100, 100], [0, 0, 0, 0],
+                                [10, 15, 20, 20], [23, 35, 70, 60]], dtype=torch.float)
+
+        box_xywh = ops.box_xyxy_to_xywh(box_tensor)
+        assert exp_xywh.size() == torch.Size([4, 4])
```
Comment (vfdev-5) [inline test/test_ops.py:660]: nit: in this project we use `self.assertEqual` and similar methods to test things... I have no strong opinion about that, but it would be better to keep conform test code...

Comment (oke-aditya) [reply test/test_ops.py:660]: K I will change the asserts for consistency 👍

Comment (pmeier) [reply test/test_ops.py:660]: @vfdev-5 Not sure I agree here. `self.assertEqual` is the `unittest` style. AFAIK we run all our tests with `pytest` and they have [great introspection](https://docs.pytest.org/en/stable/assert.html). While the "old" style is fully compatible, I don't think we should enforce it, but rather embrace the `pytest` way.

File: test/test_ops.py [hunk 7]
```diff
@@ -655,7 +695,7 @@ def test_box_area(self):
         calc_area = ops.box_area(box_tensor)
         assert calc_area.size() == torch.Size([2])
         assert calc_area.dtype == box_tensor.dtype
-        assert torch.all(torch.eq(calc_area, expected)).item() is True
```
Comment (pmeier) [inline test/test_ops.py:658]: Lets keep the scope of this PR narrow and fix this in a follow up PR. The same applies to the other changes not related to box conversion.

File: test/test_ops.py (hunk 6 above)
Comment (oke-aditya) [reply test/test_ops.py:660]: Let us decide and fix up this in a follow up PR. Right now tests are passing.

File: test/test_ops.py (hunk 7 above)
Comment (oke-aditya) [reply test/test_ops.py:658]: Ok let me undo this and fix in next one 👍

File: test/test_ops.py (hunk 6 above)
Comment (pmeier) [reply test/test_ops.py:660]: This is more of a general decision and has little to do with this PR anyway.

File: torchvision/ops/boxes.py (hunk 3 above)
Comment (pmeier) [reply torchvision/ops/boxes.py:138]: Basically, we need to define what `cx`, `cy`, and so on are. For example: "`(cx, cy)` represents the center of the bounding box".

File: test/test_ops.py (hunk 1 above)
Comment (pmeier) [reply test/test_ops.py]: @oke-aditya Tip for the next time: if you fully agree with a suggestion, i.e. do not want to include any other changes, you can merge it directly from the GitHub UI. If you navigate to the "Files changed" tab, you can select multiple suggestions and commit them as a batch.

Comment (oke-aditya) [reply test/test_ops.py]: Good idea, 👍 Will try that next time. I prefer local git as I am a bit in control of what I'm doing.

File: test/test_ops.py (hunk 6 above)
Comment (oke-aditya) [reply test/test_ops.py:660]: In a follow up PR Once we decide on what `assert` should be used, I will replace all of them in completely.

File: test/test_ops.py (hunk 7 above)
Comment (oke-aditya) [reply test/test_ops.py:658]: Reverted assertion.

File: test/test_ops.py (hunk 6 above)
Comment (fmassa) [reply test/test_ops.py:660]: Let's use `assertEqual`. We moved away from `assert` in the past, see https://github.com/pytorch/vision/issues/1483

File: torchvision/ops/boxes.py [hunk 8]
```diff
@@ -133,6 +133,78 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
...
+    Converts bounding boxes from (x, y, w, h) format to (x1, y1, x2, y2) format.
+
+    Arguments:
+        boxes (Tensor[N, 4]): boxes in (x, y, w, h) which will be converted.
+
+    Returns:
+        boxes (Tensor[N, 4]): boxes in (x1, y1, x2, y2) format.
+    """
+    boxes[:, 2] = boxes[:, 0] + boxes[:, 2]  # x + w
+    boxes[:, 3] = boxes[:, 1] + boxes[:, 3]  # y + h
```
Comment (fmassa) [inline torchvision/ops/boxes.py:189]: This is an in-place operation, and can be misleading for users who didn't expect this op to change their original boxes.
I propose we perform this out-of-place, via `stack`
```suggestion
    x, y, w, h = boxes.unbind(-1)
    return torch.stack([x, y, x + w, y + h], dim=-1)
```

File: torchvision/ops/boxes.py [hunk 9]
```diff
@@ -133,6 +133,78 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
...
+    Converts bounding boxes from (x1, y1, x2, y2) format to (x, y, w, h) format.
+
+    Arguments:
+        boxes (Tensor[N, 4]): boxes in (x1, y1, x2, y2) which will be converted.
+
+    Returns:
+        boxes (Tensor[N, 4]): boxes in (x, y, w, h) format.
+    """
+    boxes[:, 2] = boxes[:, 2] - boxes[:, 0]  # x2 - x1
+    boxes[:, 3] = boxes[:, 3] - boxes[:, 1]  # y2 - y1
```
Comment (fmassa) [inline torchvision/ops/boxes.py:204]: Same comment here about in-place operation

File: torchvision/ops/boxes.py (hunk 3 above)
Comment (fmassa) [reply torchvision/ops/boxes.py:138]: Totally agree with @pmeier here, it would be great to be a bit more explicit on what this represents

File: torchvision/ops/boxes.py [hunk 10]
```diff
@@ -133,6 +133,78 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
...
+        boxes (Tensor[N, 4]): boxes in (cx, cy, w, h) format which will be converted.
+
+    Returns:
+        boxes (Tensor(N, 4)): boxes in (x1, y1, x2, y2) format.
+    """
+    # We need to change all 4 of them so some temporary variable is needed.
+    x1 = boxes[:, 0] - (0.5 * boxes[:, 2])
+    y1 = boxes[:, 1] - (0.5 * boxes[:, 3])
+    x2 = boxes[:, 0] + (0.5 * boxes[:, 2])
+    y2 = boxes[:, 1] + (0.5 * boxes[:, 3])
```
Comment (fmassa) [inline torchvision/ops/boxes.py:150]: For less typing an a bit clearer reading, I propose we unpack first to different coordinates:
```suggestion
    cx, cy, w, h = boxes.unbind(-1)
    x1 = cx - 0.5 * w
    y1 = cy - 0.5 * h
    x2 = cx + 0.5 * w
    y2 = cy + 0.5 * h
```

**Review by fmassa — CHANGES_REQUESTED** [review]
Thanks a lot for the PR @oke-aditya and for the reviews @pmeier and @vfdev-5 !

I have a couple of comments, most notably that we should avoid in-place operations in the input argument, I already had subtle bugs in the past because of this (imagine your results after the 1st epoch being completely wrong).

I also have a meta question that I would like to discuss here: we originally discussed about adding 2 conversion functions (`xyxy_to_xywh` and the other way around), but we also added the `cxcywh_to_xyxy` variants as well.
This brings the question of the scalability of the approach, as for each new format it adds at least 2 new functions (or ~ `2 * (n ** 2 - (n - 1) ** 2)` if we do the full conversion matrix).
For reference, [Detectron2 uses a `BoxMode`](https://github.com/facebookresearch/detectron2/blob/cb7e86f821f0bedb3da8f1e0a527dd821d8e5f7e/detectron2/structures/boxes.py#L13-L130) class to represent / implement the conversion types, which lets it handle the conversions as it wishes, with only a single entry-point.

I'm not advocating for using `BoxMode` (or something like this), but my original idea was that we would only be adding support for `xyxy` and `xywh`, which is still manageable.

I'm looking forward to your thoughts

**Comment by pmeier** [comment]
@fmassa 

> This brings the question of the scalability of the approach, as for each new format it adds at least 2 new functions (or ~ `2 * (n ** 2 - (n - 1) ** 2)` if we do the full conversion matrix).

Either you have a typo in your equation or something is off. It reduces down to `2 * (2*n - 1)` and it should grow quadratically. IMO the number of functions for the full conversion matrix should be `n * (n - 1)`.

> I'm not advocating for using `BoxMode` (or something like this), but my original idea was that we would only be adding support for `xyxy` and `xywh`, which is still manageable.
> 
> I'm looking forward to your thoughts

Speaking form a position of ignorance as I've not worked with object detection very often: do we even need to scale this? I mean are there more than the three representations (bottom left with height and width / center with height and width / bottom left and top right) at all? Sure, it is possible that any corner might be used as anchor as well as any two opposite corners might be used, but is that common? If that is the case, we should discuss which variants we want to support and how elaborate the support should be.

If not I think we can implement 2 functions for each representation and be done with it as we are a general vision library rather than specialized for object detection.

File: test/test_ops.py (hunk 6 above)
Comment (oke-aditya) [reply test/test_ops.py]: Changed to assertEqual. In a new PR I will change in other places as well.

**Comment by oke-aditya** [comment]
I'm not highly experienced or qualified but would like to share some thoughts (ignore if they make no sense).

- It's not wise to have all the conversions as the matrix is quadratic and once we provide it we have to keep providing.

- Probably cascading these operations gets the job done for now, so we have 2 conversions for each sets. I guess there are not many `popular` ways of representing boxes, I have seen just these three (correct me if I'm wrong).

- If users requests to provide more such methods then we can think in future! Converting a representation to `xyxy` format is something we can provide easily and maintain.

Let me know thoughts, let us not create a feature that is not maintainable by us.

File: torchvision/ops/boxes.py (hunk 3 above)
Comment (oke-aditya) [reply torchvision/ops/boxes.py:138]: I added them, but I'm slightly bad at documenting. Let me know if there needs changes.

File: torchvision/ops/boxes.py [hunk 11]
```diff
@@ -133,6 +133,85 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
...
+    Arguments:
+        boxes (Tensor[N, 4]): boxes in (x, y, w, h) which will be converted.
+
+    Returns:
+        boxes (Tensor[N, 4]): boxes in (x1, y1, x2, y2) format.
+    """
+    x, y, w, h = boxes.unbind(-1)
+    return torch.stack([x, y, x + w, y + h], dim=-1)
+    return boxes
+
```
Comment (fmassa) [inline torchvision/ops/boxes.py:195]:
```suggestion

```

**Comment by fmassa** [comment]
> Either you have a typo in your equation or something is off. It reduces down to 2 * (2*n - 1) and it should grow quadratically. IMO the number of functions for the full conversion matrix should be n * (n - 1).

My intent was to say how many more functions we would need to add if we were to go from `n-1` to `n` modes.

> Speaking form a position of ignorance as I've not worked with object detection very often: do we even need to scale this? I mean are there more than the three representations (bottom left with height and width / center with height and width / bottom left and top right) at all? Sure, it is possible that any corner might be used as anchor as well as any two opposite corners might be used, but is that common? If that is the case, we should discuss which variants we want to support and how elaborate the support should be.

We don't need to add all possible conversion combinations, but just by the fact that we are already adding 4 new functions makes me think that this approach doesn't scale. I'm ok to always have the conversions passing through `xyxy`, but still even in this case we already have a lot of functions.

Let me illustrate with another examples on why I think we should follow a different approach here:
- [OpenCV allow to convert images between different colorspaces](https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_imgproc/py_colorspaces/py_colorspaces.html), but it only provides one function for that, which accepts a second argument (flag) that indicates the direction of the conversion
- PIL as well, except that it already knows the colorspace of the current image (because it's stored internally)
- Detectron2 has the `BoxMode` class for abstracting this away
- [albumentations uses a single function for defining how to convert boxes](https://github.com/albumentations-team/albumentations/blob/97b640fe1fcdc6d991a86bd44ec56cf3c02aae4e/albumentations/augmentations/bbox_utils.py#L195-L252), via a `source_format` argument (the `target_format` is always the same)

Here is my proposal: implement a function called `convert_box(boxes, input_fmt, output_fmt)` (better names welcome!) which is the single entry-point for performing those conversions. This way, the user only needs to care about one function.
`input_fmt` and `output_fmt` can be strings such as `xyxy` and `xywh`, so that we don't need to introduce any new abstractions.

Thoughts?

**Comment by oke-aditya** [comment]
Great thoughts. I guess it makes much more sense and generic. User can simply pass 2 strings and get his bounding boxes converted without thinking much. So user side it is less headache.

Coming to our side, if we provide such a function. We would need to provide all conversions as the user will not know which methods are possible, he would simply expect that the boxes should be converted!

All conversions can occur through `xyxy` internally, that is inefficient but still it would reduce our codebase and we would have less to maintain. 
We can add further efficient operations without affecting the API. Internally we might have lot of functions but we expose just 1 to user.

I completely agree with this opinion, really a good idea (probably we should have discussed in issue more and I did hurry in jumping to code, sorry for that)

**Comment by fmassa** [comment]
> Coming to our side, if we provide such a function. We would need to provide all conversions as the user will not know which methods are possible, he would simply expect that the boxes should be converted!

I would go with the approach you mentioned just afterwards -- always go through `xyxy` as an intermediate representation.
Here is some pseudo-code illustrating one potential implementation:
```
def convert_boxes(boxes, in_fmt, out_fmt):
    allowed_fmts = ...
    assert in_fmt in allowed_fmts
    assert out_fmt in allowed_fmts
    if in_fmt == out_fmt:
        return boxes.clone()  # to ensure always returning a copy
    if in_fmt != 'xyxy' and out_fmt != 'xyxy':
        # convert one to xyxy and change either in_fmt or out_fmt to xyxy
    # dispatch to the existing functions
    ...
```

Also, I think it might be preferable to spell it as `convert_boxes` instead of `convert_box` because it supports multiple boxes at once, but I think we already named it `box_area` in the past so maybe it's not that much of an issue?

**Comment by oke-aditya** [comment]
We can name to `convert_box` it won't be an issue as we always expect `box` to be `Tensor[N]` E.g. `box_area`, `box_iou` so it stays consistent.

Right now I think I will refactor to work internally with `xyxy` and the prototype seems great to me. 

Let me refactor the code. 

Should I do in a new PR or continue here? this PR will become quite dirty. 
Reason being I will need to change code, docs and tests. Though a lot of it is re-use.

I guess all these conversion functions used internally be named as `_box_xyxy_to_xywh` since these will be used internally and we won't provide docs as well as include them in `__init__` and `__all__ `.

This will effectively save a lot of efforts in maintaining docs for these functions internally and just maintain clear documentation and usage of the above `convert_box` function.

**Comment by fmassa** [comment]
We can continue in this PR, and the proposal of renaming the current functions as `_box_xyxy_to_xywh` is what I would have suggested as well.

Good point about `box_area` / `box_iou`, which makes me think that we should maybe name it as `box_convert` maybe?

Don't worry about the history of the commits, it will all get squashed by GitHub before merging.

One thing to keep in mind: can you add a test for torchscriptability as well? Something like
```python
out = box_convert(boxes, 'xyxy', 'xywh')
scripted_fn = torch.jit.script(box_convert)
out_script = scripted_fn(boxes, 'xyxy', 'xywh')
self.assertTrue((out - out_script).abs().max() < TOLERANCE)
```
This will ensure that our transform is ready to be exported to C++. Let us know if you have issues making the code work with torchscript.

And thanks a lot for your help!

File: torchvision/ops/boxes.py (hunk 3 above)
Comment (oke-aditya) [reply torchvision/ops/boxes.py]: Marking as resolved due to new API.

File: torchvision/ops/boxes.py (hunk 11 above)
Comment (oke-aditya) [reply torchvision/ops/boxes.py]: Resolved 👍

**Comment by oke-aditya** [comment]
Sorry for the delay. 
As per discussion, I rearranged the code a bit. I made function `box_convert` in `boxes.py` as per discussion above.

The other utility conversion functions which are few for now (might grow in future) I shifted to separate file `_box_convert.py`. All renamed according to conventions. Let me know if this is fine, they would pollute the `boxes.py` file and occupy too much unnecessary code which is abstracted out hence I shifted it.

Simply refactored the tests for this new API, I will add the JIT test and one more conversion test as well.

Documentation is only generated for `box_convert` function, not for others.

Let me know if this works and if it needs changes :-)

**Comment by oke-aditya** [comment]
I added the JIT test as well, but it somehow kept failing for me locally, I'm not sure about it. Can someone have a look, please?

**Comment by oke-aditya** [comment]
> Closes #2687
> 
> - [x] Added code
> - [x] Added documentation
> - [x] Added tests
> 
> I have added as per the Issue, two utility function to convert boxes to pascal VOC format (x1 y1 x2 y2).
> 
> The tests convert boxes to other format and vice-versa. This ensures that both operations are identical and can be interchangeably used. Tests passed locally. 
> 
> This is ready for review. Do let me know!
> 
> cc @pmeier

File: torchvision/ops/boxes.py [hunk 12]
```diff
@@ -133,6 +134,54 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
     return clipped_boxes.reshape(boxes.shape)
 
 
+def box_convert(boxes: Tensor, in_fmt: str, out_fmt: str) -> Tensor:
+    """
+    Converts boxes from given in_fmt to out_fmt.
+
```
Comment (fmassa) [inline torchvision/ops/boxes.py:140]: Can we specify in the documentation string what are the possible supported formats, as well as what they represent?
Something like:
```
Supported formats are:
    'xyxy': boxes are represented via their corners, with x1, y1 the top left corner and ...
```

File: torchvision/ops/boxes.py [hunk 13]
```diff
@@ -133,6 +134,54 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
...
+    Returns:
+        boxes (Tensor[N, 4]): Boxes into converted format.
+    """
+    allowed_fmts = ("xyxy", "xywh", "cxcywh")
+    assert in_fmt in allowed_fmts
+    assert out_fmt in allowed_fmts
+
+    if in_fmt == out_fmt:
+        # return boxes.clone()  # to ensure always returning a copy
```
Comment (fmassa) [inline torchvision/ops/boxes.py:154]:
```suggestion
        # ensure we always return a copy
```

File: torchvision/ops/boxes.py [hunk 14]
```diff
@@ -133,6 +134,54 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
...
+        boxes_converted = boxes.clone()
+        return boxes_converted
+
+    else:
```
Comment (fmassa) [inline torchvision/ops/boxes.py:158]: Given that we already returned before, we can remove this `else` case and save one level of identation

File: torchvision/ops/boxes.py [hunk 15]
```diff
@@ -133,6 +134,54 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
...
+        if in_fmt != 'xyxy' and out_fmt != 'xyxy':
+            if in_fmt == "xywh":
+                boxes_xyxy = _box_xywh_to_xyxy(boxes)
+                if out_fmt == "cxcywh":
+                    boxes_converted = _box_xyxy_to_cxcywh(boxes_xyxy)
+
+            elif in_fmt == "cxcywh":
+                boxes_xyxy = _box_cxcywh_to_xyxy(boxes)
+                if out_fmt == "xywh":
+                    boxes_converted = _box_xyxy_to_xywh(boxes_xyxy)
```
Comment (fmassa) [inline torchvision/ops/boxes.py:168]: While this is fine, my first thought was to so something like the following
```python
            if in_fmt == "xywh":
                boxes = _box_xywh_to_xyxy(boxes)
                in_fmt = "xyxy"
            elif in_fmt == "cxcywh":
                boxes = _box_cxcywh_to_xyxy(boxes)
                in_fmt = "xyxy"
```
and let the rest of the dispatch to be done in the last branch. This way, we don't need to replicate the `out` dispatch logic here.

You don't need to change this here so that we can move forward quickly with this PR, but it would be good to send a follow-up PR improving this part after this PR gets merged. Thoughts?

File: torchvision/ops/boxes.py [hunk 16]
```diff
@@ -133,6 +134,54 @@ def clip_boxes_to_image(boxes: Tensor, size: Tuple[int, int]) -> Tensor:
...
 
 
+def box_convert(boxes: Tensor, in_fmt: str, out_fmt: str) -> Tensor:
+    """
+    Converts boxes from given in_fmt to out_fmt.
+
+    Arguments:
+        boxes (Tensor[N, 4]): boxes which will be converted.
+        in_fmt (str): Input format of given boxes.
+        out_fmt (str): Output format of given boxes.
```
Comment (fmassa) [inline torchvision/ops/boxes.py:144]: Same here, can you also specify the list of supported formats? Like `supported formats are ['xyxy', 'xywh', 'cxcywh']`

File: test/test_ops.py [hunk 17]
```diff
@@ -647,6 +647,102 @@ def test_convert_boxes_to_roi_format(self):
...
+
+        # Reverse conversion
+        box_xywh = ops.box_convert(box_cxcywh, in_fmt="cxcywh", out_fmt="xywh")
+        self.assertEqual(box_xywh.size(), torch.Size([4, 4]))
+        self.assertEqual(box_xywh.dtype, box_tensor.dtype)
+        assert torch.all(torch.eq(box_xywh, box_tensor)).item()
+
+    # def test_bbox_convert_jit(self):
+    #     box_tensor = torch.tensor([[0, 0, 100, 100], [0, 0, 0, 0],
+    #                               [10, 15, 30, 35], [23, 35, 93, 95]], dtype=torch.float)
```
Comment (fmassa) [inline test/test_ops.py:732]: Two options here:
- we merge the PR now and try to fix torchscript later
- we fix torchscript right now.

Do you remember what type of errors you were facing? I'm fine with both approaches, so that we can move forward with this PR (but we should fix torchscript soon if we merge this without torchscript support)

**Review by fmassa — COMMENTED** [review]
The code looks great, thanks a lot for all your work @oke-aditya !

Do you remember what was the test failure that you were facing with torchscript? From looking at your implementation I don't see why it should fail.

I only have a couple of documentation suggestions, the other comment can be left for a future PR

**Comment by oke-aditya** [comment]
I will add these documentation fixes.
I'm not quite experienced with torchscript and let us fix in a new follow-up PR. I will open it as soon as this gets merged. 

(I guess all the operations in `boxes` have support for torchscript and let's keep it in October release as wll)

IIRC torchscript failed due to not finding some else block for if (I will post the error stack in the new PR)

The code can be cleaned up as you suggested, but I would like to have torchscript support first and then clean up. Reason being this code is not quite efficient in conversions.

File: test/test_ops.py [hunk 18]
```diff
@@ -672,7 +768,7 @@ def test_iou(self):
         # Check if all elements of tensor are as expected.
         assert out.size() == torch.Size([3, 3])
         tolerance = 1e-4
-        assert ((out - expected).abs().max() < tolerance).item() is True
+        assert ((out - expected).abs().max() < tolerance).item()
```
Comment (fmassa) [inline test/test_ops.py:771]: Reverting this change as suggested by @pmeier in https://github.com/pytorch/vision/pull/2710#discussion_r495760308 . Let's do this in a separate PR
```suggestion
        assert ((out - expected).abs().max() < tolerance).item() is True
```

File: test/test_ops.py [hunk 19]
```diff
@@ -690,7 +786,7 @@ def test_gen_iou(self):
...
```
Comment (fmassa) [inline test/test_ops.py:789]: Reverting as suggested by @pmeier in https://github.com/pytorch/vision/pull/2710#discussion_r495760308
```suggestion
        assert ((out - expected).abs().max() < tolerance).item() is True
```

File: torchvision/ops/boxes.py (hunk 12 above)
Comment (oke-aditya) [reply torchvision/ops/boxes.py:140]: added 👍

File: torchvision/ops/boxes.py (hunk 13 above)
Comment (oke-aditya) [reply torchvision/ops/boxes.py:154]: Removed the comment 👍

File: torchvision/ops/boxes.py (hunk 14 above)
Comment (oke-aditya) [reply torchvision/ops/boxes.py:158]: I added this in current PR, lets leave next one for torchscript.

File: torchvision/ops/boxes.py (hunk 16 above)
Comment (oke-aditya) [reply torchvision/ops/boxes.py:144]: Added 👍

File: torchvision/ops/boxes.py (hunk 15 above)
Comment (oke-aditya) [reply torchvision/ops/boxes.py:168]: I guess I will leave this for next PR, my original idea was we support direct conversions at some point of time and it should be simpler to refactor for future. But this too works fine.

File: test/test_ops.py (hunk 18 above)
Comment (oke-aditya) [reply test/test_ops.py]: Reverted

File: test/test_ops.py (hunk 19 above)
Comment (oke-aditya) [reply test/test_ops.py]: Reverted

**Comment by oke-aditya** [comment]
I guess in one of the follow up PRs, I will clean all the `assert` statements, there are many places I could see such inconsistent use.

**Review by fmassa — APPROVED** [review]
Sounds good, let's fix torchscript and the minor refactorings in a follow-up PR.

Thanks a lot @oke-aditya !

**Comment by oke-aditya** [comment]
Just let me add documentation. I'm about to push changes 😅

File: torchvision/ops/_box_convert.py [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+    (x2, y2) refer to bottom right of bounding box
+    Arguments:
+        boxes (Tensor[N, 4]): boxes in (x1, y1, x2, y2) which will be converted.
+
+    Returns:
+        boxes (Tensor[N, 4]): boxes in (x, y, w, h) format.
+    """
+    x1, y1, x2, y2 = boxes.unbind(-1)
+    x2 = x2 - x1  # x2 - x1
+    y2 = y2 - y1  # y2 - y1
```
Comment (fmassa) [inline torchvision/ops/_box_convert.py:81]: For a future PR: can you rename this to be `w` and `h`? This now is a bit misleading

**Review by fmassa — APPROVED** [review]
Thanks a lot! Looking forward to the torchscript improvements!

File: torchvision/ops/_box_convert.py (hunk 20 above)
Comment (oke-aditya) [reply torchvision/ops/_box_convert.py:81]: Sure 👍
